code/game/machinery/shieldgen.dm

```diff
@@ -305,142 +305,113 @@
 
-	active = !active
-	user.visible_message("[user] turned the shield generator [active ? "on" : "off"].", \
-	"You turn [active ? "on" : "off"] the shield generator.", \
-	"You hear heavy droning [active ? "" : "fade out"].")
-	if(active)
-		for(var/dir in list(NORTH, SOUTH, EAST, WEST))
-			cleanup(dir)
+	if(!activated)
+		activate()
+	else
+		deactivate()
+
+	user.visible_message("<span class='notice'>[user] turned the shield generator [activated ? "on" : "off"].</span>", \
+		"<span class='notice'>You turn [activated ? "on" : "off"] the shield generator.</span>", \
+		"You hear heavy droning [activated ? "" : "fade out"].")
+
 	update_icon(UPDATE_ICON_STATE)
 	add_fingerprint(user)
 
 /obj/machinery/shieldwallgen/process()
-	spawn(100)
-		power()
-		if(power)
-			storedpower -= 50 //this way it can survive longer and survive at all
-	if(storedpower >= maxstoredpower)
-		storedpower = maxstoredpower
-	if(storedpower <= 0)
-		storedpower = 0
-
-	if(active)
-		if(state)
-			active = FALSE
-			return
-		spawn(1)
-			setup_field(1)
-		spawn(2)
-			setup_field(2)
-		spawn(3)
-			setup_field(4)
-		spawn(4)
-			setup_field(8)
-	if(active)
-		if(!power)
-			visible_message("<span class='warning'>[name] shuts down due to lack of power!</span>", \
+	if(!activated)
+		return
+	if(!try_charge_shields_power())
+		visible_message("<span class='warning'>[name] shuts down due to lack of power!</span>", \
 				"You hear heavy droning fade out")
-			active = FALSE
-			for(var/dir in list(NORTH, SOUTH, EAST, WEST))
-				cleanup(dir)
-			update_icon(UPDATE_ICON_STATE)
+		deactivate()
+		update_icon(UPDATE_ICON_STATE)
 
-/obj/machinery/shieldwallgen/proc/setup_field(NSEW = 0)
-	var/turf/T = loc
-	var/turf/T2 = loc
-	var/obj/machinery/shieldwallgen/G
-	var/steps = 0
-	var/oNSEW = 0
 
-	if(!NSEW)//Make sure its ran right
-		return
+/obj/machinery/shieldwallgen/proc/activate()
+	activated = TRUE
+	for(var/direction in GLOB.cardinal)
+		INVOKE_ASYNC(src, .proc/try_link_generators, direction)
 
-	if(NSEW == 1)
-		oNSEW = 2
-	else if(NSEW == 2)
-		oNSEW = 1
-	else if(NSEW == 4)
-		oNSEW = 8
-	else if(NSEW == 8)
-		oNSEW = 4
-
-	for(var/dist = 0, dist <= 9, dist += 1) // checks out to 8 tiles away for another generator
-		T = get_step(T2, NSEW)
-		T2 = T
-		steps += 1
-		if(locate(/obj/machinery/shieldwallgen) in T)
-			G = (locate(/obj/machinery/shieldwallgen) in T)
-			steps -= 1
-			if(!G.active)
-				return
-			G.cleanup(oNSEW)
+/obj/machinery/shieldwallgen/proc/try_link_generators(direction)
+	var/turf/current_turf = loc
+	var/obj/machinery/shieldwallgen/other_generator
+	var/list/traveled_turfs = list()
+	for(var/distance in 1 to 9)
+		current_turf = get_step(current_turf, direction)
+		other_generator = locate(/obj/machinery/shieldwallgen) in current_turf
+		if(other_generator)
 			break
+		traveled_turfs += current_turf
 
-	if(isnull(G))
+	if(!other_generator?.activated)
 		return
 
-	T2 = loc
+	sleep(rand(1, 4)) // TODO check if we want this behaviour
```

Comment:
I'm fine with removing this one if the design team doesn't mind that there is no delay on starting up the generators